app/services/scholarsphere_location_operator.rb

```diff
@@ -0,0 +1,86 @@
+# frozen_string_literal: true
+
+class ScholarsphereLocationOperator
+  def initialize(params:, token:)
+    @params = params
+    @token = token
+  end
+
+  def perform
+    if !@token.write_access?
+      operator_response(401, 'unauthorized')
+    elsif !params_valid?
+      operator_response(422, 'params_invalid')
+    elsif update_open_access_location.present?
+      operator_response(200, 'success')
+    elsif @existing_location
+      operator_response(422, 'existing_location')
+    else
+      operator_response(404, 'no_content')
+    end
+  end
+
+  private
+
+    attr_reader :params, :token
+
+    def params_valid?
+      return false if params[:scholarsphere_open_access_url].blank?
+
+      !(params[:activity_insight_id].present? && params[:doi].present?)
+    end
+
+    def update_open_access_location
+      locations = []
+
+      ActiveRecord::Base.transaction do
+        filtered_publications.each do |publication|
+          return if existing_location?(publication)
+
+          location = find_or_create_location(publication)
+          locations << location.url if location.valid?
+        end
+      end
+
+      locations
+    end
+
+    def filtered_publications
+      publications = []
+
+      if params[:activity_insight_id].present?
+        publications = Publication::filter_by_activity_insight_id(Publication, params[:activity_insight_id])
+      elsif params[:doi].present?
+        publications = Publication::filter_by_doi(Publication, params[:doi])
+      end
+
+      publications
+    end
+
+    def existing_location?(publication)
+      @existing_location ||= publication.open_access_locations
+        .filter { |l| l.source == Source::USER }
+        .index_by(&:url)
+        .key?(params[:scholarsphere_open_access_url])
+    end
+
+    def find_or_create_location(publication)
+      publication.open_access_locations.find_or_create_by(
+        source: Source::USER,
+        url: params[:scholarsphere_open_access_url]
+      )
+    end
+
+    def operator_response(code, message)
+      OperatorResponse.new(code: code, message: message)
+    end
+
+    class OperatorResponse
```

Comment:
This can just be a struct, but I don't think there's any disadvantage to the way you have it here.

Reply:
Interesting, I started out with a struct then turned it into a class when I thought I would add couple other methods but eventually I did not need those additinal methods so yeah it can actually be just a Struct. I will change it.

Reply:
@awead I changed it to an OpenStruct.